core/services/ccip/orm.go

```diff
@@ -0,0 +1,163 @@
+
+	ClearGasPricesByDestChain(ctx context.Context, destChainSelector uint64, to time.Time) error
+	ClearTokenPricesByDestChain(ctx context.Context, destChainSelector uint64, to time.Time) error
+}
+
+type orm struct {
+	ds sqlutil.DataSource
+}
+
+var _ ORM = (*orm)(nil)
+
+func NewORM(ds sqlutil.DataSource) (ORM, error) {
+	if ds == nil {
+		return nil, fmt.Errorf("datasource to CCIP NewORM cannot be nil")
+	}
+
+	return &orm{
+		ds: ds,
+	}, nil
+}
+
+func (o *orm) GetGasPricesByDestChain(ctx context.Context, destChainSelector uint64) ([]GasPrice, error) {
+	var gasPrices []GasPrice
+	stmt := `
+		SELECT DISTINCT ON (source_chain_selector)
+		source_chain_selector, gas_price, created_at
+		FROM ccip.observed_gas_prices
+		WHERE chain_selector = $1
+		ORDER BY source_chain_selector, created_at DESC;
+	`
+	err := o.ds.SelectContext(ctx, &gasPrices, stmt, destChainSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	return gasPrices, nil
+}
+
+func (o *orm) GetTokenPricesByDestChain(ctx context.Context, destChainSelector uint64) ([]TokenPrice, error) {
+	var tokenPrices []TokenPrice
+	stmt := `
+		SELECT DISTINCT ON (token_addr)
+		token_addr, token_price, created_at
+		FROM ccip.observed_token_prices
+		WHERE chain_selector = $1
+		ORDER BY token_addr, created_at DESC;
+	`
+	err := o.ds.SelectContext(ctx, &tokenPrices, stmt, destChainSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	return tokenPrices, nil
+}
+
+func (o *orm) InsertGasPricesForDestChain(ctx context.Context, destChainSelector uint64, jobId int32, gasPrices []GasPriceUpdate) error {
+	if len(gasPrices) == 0 {
+		return nil
+	}
+
+	now := time.Now()
+	insertData := make([]map[string]interface{}, 0, len(gasPrices))
+	for _, price := range gasPrices {
+		insertData = append(insertData, map[string]interface{}{
+			"chain_selector":        destChainSelector,
+			"job_id":                jobId,
+			"source_chain_selector": price.SourceChainSelector,
+			"gas_price":             price.GasPrice,
+			"created_at":            now,
+		})
+	}
+
+	stmt := `INSERT INTO ccip.observed_gas_prices (chain_selector, job_id, source_chain_selector, gas_price, created_at)
+		VALUES (:chain_selector, :job_id, :source_chain_selector, :gas_price, :created_at);`
+	_, err := o.ds.NamedExecContext(ctx, stmt, insertData)
+	if err != nil {
+		err = fmt.Errorf("error inserting gas prices for job %d: %w", jobId, err)
+	}
+
+	return err
+}
+
+func (o *orm) InsertTokenPricesForDestChain(ctx context.Context, destChainSelector uint64, jobId int32, tokenPrices []TokenPriceUpdate) error {
+	if len(tokenPrices) == 0 {
+		return nil
+	}
+
+	now := time.Now()
+	insertData := make([]map[string]interface{}, 0, len(tokenPrices))
+	for _, price := range tokenPrices {
+		insertData = append(insertData, map[string]interface{}{
+			"chain_selector": destChainSelector,
+			"job_id":         jobId,
+			"token_addr":     price.TokenAddr,
+			"token_price":    price.TokenPrice,
+			"created_at":     now,
```

Comment:
I think we typically let the database set this field instead of passing it explicitly.

Reply:
my understanding is there isn't a single best practice when it comes to passing in timestamp v.s DB default, in this case DB default makes tests tricky, the get logic is dependent on sorting by timestamp, timestamp during tests is not increasing reliably

Reply:
I think of best practice as using db clock for both inserts and deletes rather than golang clock.

`now()` is an alias for `transaction_timestamp()` which returns the start time of the current transaction.  That does have issues with tests, because we run all tests within a single transaction.  However, I think any of the following should probably work without issues: `statement_timestamp()`, `clock_timestamp()`, or just `TIMESTAMP 'now'` https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-CURRENT